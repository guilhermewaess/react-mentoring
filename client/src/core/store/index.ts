import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import logger from 'redux-logger';
import { persistReducer, persistStore } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
// import { Promise } from 'redux-promise';
import { reducers } from './reducers';

// workaround for lack of type file
const reduxPromise = require('redux-promise').default; //tslint:disable-line


declare module 'redux' {
  export type GenericStoreEnhancer = any;
}

const persistConfig = {
  key: 'state',
  storage,
};
console.log(reduxPromise); //tslint:disable-line
const persistedReducer = persistReducer(persistConfig, reducers);

const store = createStore(
  persistedReducer,
  {},
  composeWithDevTools(applyMiddleware(reduxPromise, logger)),
);
const persistor = persistStore(store);

export { store, persistor };
