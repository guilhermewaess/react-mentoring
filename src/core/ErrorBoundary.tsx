import * as React from 'react';

interface IState {
  hasError: boolean;
}

export default class ErrorBoundary extends React.Component<any, IState> {
  public state: IState;

  constructor(props: any) {
    super(props);
    this.state = { hasError: false };
  }

  public componentDidCatch(error: any, errorInfo: any) {
    this.setState({ hasError: true });
  }

  public render() {
    if (this.state.hasError) {
      return <h1>Ops</h1>;
    }

    return this.props.children;
  }
}
