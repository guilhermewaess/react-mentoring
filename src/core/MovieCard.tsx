import * as React from 'react';
import { NavLink } from 'react-router-dom';
import {
  Badge,
  Card,
  CardBody,
  CardImg,
  CardSubtitle,
  CardTitle,
} from 'reactstrap';
import './MovieCard.scss';

const MovieGenres = ({ genres }: { genres: string[] }) => (
  <div>
    {genres.map((genre, index) => <Badge key={index} color="secondary mr-1">{genre}</Badge>)}
  </div>
);

export default ({ match, movie }: any) => (
  <div className="movie-card">
    <NavLink to={`movie-details/${movie.id}`}>
      <Card>
        <CardImg
          top={true}
          width="100%"
          height="400px"
          src={movie.poster_path}
          alt="Movie Image"
        />
        <CardBody>
          <CardTitle>{movie.title}</CardTitle>
          <CardSubtitle>
            <MovieGenres genres={movie.genres} />
          </CardSubtitle>
        </CardBody>
      </Card>
    </NavLink>
  </div>
);
