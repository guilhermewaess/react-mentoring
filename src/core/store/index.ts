import { applyMiddleware, createStore } from 'redux';
import logger from 'redux-logger';
import { persistReducer, persistStore } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import * as reduxPromise from 'redux-promise';
import { reducers } from './reducers';

const persistConfig = {
  key: 'state',
  storage,
}

const persistedReducer = persistReducer(persistConfig, reducers)

const store = createStore(
  persistedReducer,
  {},
  applyMiddleware(logger, reduxPromise)
)
const persistor = persistStore(store)

export {
  store, persistor
}
