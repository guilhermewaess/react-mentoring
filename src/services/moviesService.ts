import axios from 'axios';
const baseUrl = 'http://react-cdp-api.herokuapp.com';

export function getAll() {
  const url = `${baseUrl}/movies`;
  return axios.get(url).then(response => response.data.data);
}