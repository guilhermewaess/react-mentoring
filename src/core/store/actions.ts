import { getAll } from './../../services/moviesService';

export enum ActionTypes {
  getMovies = '[Store] - Get Movies',
}

export function getMovies() {
  return {
    type: ActionTypes.getMovies,
    payload: getAll()
  }
}
