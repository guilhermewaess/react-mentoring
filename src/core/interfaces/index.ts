import { ActionTypes } from "../store/actions";

export interface IMovie {
  budget: number;
  genres: string[];
  id: number;
  overview: string;
  poster_path: string;
  release_date: string;
  revenue: number;
  runtime: null;
  tagline: string;
  title: string;
  vote_average: number;
  vote_count: number;
}

export interface ICommonState {
  movie: IMovie;
}

export interface ICommonProps {
  movie: IMovie;
}

export interface IAction {
  type: ActionTypes,
  payload: any
}