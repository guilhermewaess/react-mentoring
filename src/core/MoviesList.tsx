import * as React from 'react';
import { Col, Container, Row } from 'reactstrap';
import { IMovie } from './interfaces';
import MovieCard from './MovieCard';

const NoFilmsFound = () => <h2>No films found</h2>;

export default ({ movies, match }: any) => {
  if (!movies.length) {
    return <NoFilmsFound />;
  }
  return (
    <Container fluid={true}>
      <Row>
        {movies.map((movie: IMovie) => (
          <Col xs={3} key={movie.id}>
            <MovieCard movie={movie} />
          </Col>
        ))}
      </Row>
    </Container>
  );
};
