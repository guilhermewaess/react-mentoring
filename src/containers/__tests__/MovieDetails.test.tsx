jest.mock('./../../core/store/actions', () => ({
  getMovies: jest.fn().mockReturnValue({
    type: 'test',
  }),
}));
jest.mock('./../../services/moviesService', () => ({
  getMovieById: jest.fn(),
}));

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import * as React from 'react';
import * as reduxPromise from 'redux-promise';
import { IMovie } from '../../core/interfaces';
import { getMovies } from './../../core/store/actions';
import { getMovieById } from './../../services/moviesService';
import MovieDetails from './../MovieDetails';
const configureStore = require('redux-mock-store'); //tslint:disable-line

describe('MovieDetailsContainer', () => {
  let component: ShallowWrapper<any, any>;
  let store: any;
  let instance: any;
  let props: any;
  let movie1: IMovie;
  let movie2: IMovie;
  let spy: jest.Mock;

  const helper = {
    createProps() {
      return {
        match: { params: { id: '100' } },
      };
    },
    createMovie(id: number) {
      return {
        budget: 10,
        genres: ['Action', 'Adventure', 'Fantasy', 'Science Fiction'],
        id,
        overview: 'We all...',
        poster_path:
          'https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg',
        release_date: '2019-04-05',
        revenue: 0,
        runtime: null,
        tagline: '',
        title: 'Shazam!',
        vote_average: 0,
        vote_count: 5,
      };
    },
  };
  beforeEach(() => {
    props = helper.createProps();
    movie1 = helper.createMovie(100);
    movie2 = helper.createMovie(200);
    spy = getMovieById as jest.Mock;
    spy.mockReset();
    spy.mockResolvedValue(movie1);
  });
  describe('when there is no movies on store', () => {
    beforeEach(() => {
      store = configureStore([reduxPromise])({ movies: [] });
      component = shallow(<MovieDetails store={store} {...props} />).dive();
      instance = component.instance();
    });
    it('should render loading', () => {
      expect(shallowToJson(component)).toMatchSnapshot();
    });
    it('should request movieById from service', () => {
      expect(getMovieById).toHaveBeenCalledWith(Number(props.match.params.id));
    });
    describe('and service resolve with success', () => {
      it('should setState with movie', async done => {
        await component.update();
        expect(component.state().movie).toEqual(movie1);
        done();
      });
      it('should get moviesByGenres', () => {
        expect(getMovies).toHaveBeenCalledWith({
          search: component.state().movie.genres[0],
          searchBy: 'genres',
          sortBy: 'release_date',
          sortOrder: 'asc',
        });
      });
    });
  });

  describe('when there is movies on store', () => {
    beforeEach(() => {
      store = configureStore([reduxPromise])({ movies: [movie1, movie2] });
      component = shallow(<MovieDetails store={store} {...props} />).dive();
      instance = component.instance();
    });
    it('should render component', () => {
      expect(shallowToJson(component)).toMatchSnapshot();
    });
    it('should not request movieById from service', () => {
      expect(getMovieById).not.toHaveBeenCalled();
    });
    it('should get movie by id from store', () => {
      component.update();
      expect(component.state().movie).toEqual(movie1);
    });
  });

  describe('when component willReceiveProps', () => {
    beforeEach(() => {
      store = configureStore([reduxPromise])({ movies: [movie1, movie2] });
      component = shallow(<MovieDetails store={store} {...props} />).dive();
      instance = component.instance();
    });
    describe('and the movieId on newProps is same', () => {
      beforeEach(() => {
        instance.componentWillReceiveProps(props);
      });
      it('should do nothing', () => {
        component.update();
        expect(component.state().movie).toEqual(movie1);
      });
    });
    describe('and the movieId on newProps is different', () => {
      beforeEach(() => {
        props.match.params.id = 200;
        instance.componentWillReceiveProps(props);
      });
      it('should get movie by id from store', () => {
        component.update();
        expect(component.state().movie).toEqual(movie2);
      });
    });
  });
});
