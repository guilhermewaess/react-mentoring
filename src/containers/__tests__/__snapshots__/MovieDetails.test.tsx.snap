// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsContainer when there is movies on store should render component 1`] = `
<div
  className="w-100 h-100"
>
  <Component
    movie={
      Object {
        "budget": 10,
        "genres": Array [
          "Action",
          "Adventure",
          "Fantasy",
          "Science Fiction",
        ],
        "id": 100,
        "overview": "We all...",
        "poster_path": "https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg",
        "release_date": "2019-04-05",
        "revenue": 0,
        "runtime": null,
        "tagline": "",
        "title": "Shazam!",
        "vote_average": 0,
        "vote_count": 5,
      }
    }
  />
  <Component
    movie={
      Object {
        "budget": 10,
        "genres": Array [
          "Action",
          "Adventure",
          "Fantasy",
          "Science Fiction",
        ],
        "id": 100,
        "overview": "We all...",
        "poster_path": "https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg",
        "release_date": "2019-04-05",
        "revenue": 0,
        "runtime": null,
        "tagline": "",
        "title": "Shazam!",
        "vote_average": 0,
        "vote_count": 5,
      }
    }
  />
  <Component
    movies={
      Array [
        Object {
          "budget": 10,
          "genres": Array [
            "Action",
            "Adventure",
            "Fantasy",
            "Science Fiction",
          ],
          "id": 100,
          "overview": "We all...",
          "poster_path": "https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg",
          "release_date": "2019-04-05",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Shazam!",
          "vote_average": 0,
          "vote_count": 5,
        },
        Object {
          "budget": 10,
          "genres": Array [
            "Action",
            "Adventure",
            "Fantasy",
            "Science Fiction",
          ],
          "id": 200,
          "overview": "We all...",
          "poster_path": "https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg",
          "release_date": "2019-04-05",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Shazam!",
          "vote_average": 0,
          "vote_count": 5,
        },
      ]
    }
  />
</div>
`;

exports[`MovieDetailsContainer when there is no movies on store should render loading 1`] = `
<div>
  Loading...
</div>
`;
